sdks/java/extensions/google-cloud-platform-core/src/test/java/org/apache/beam/sdk/extensions/gcp/util/RetryHttpRequestInitializerTest.java

```diff
@@ -209,23 +231,30 @@ public void testThrowIOException() throws IOException {
     verify(mockLowLevelRequest, times(2)).setTimeout(anyInt(), anyInt());
     verify(mockLowLevelRequest, times(2)).setWriteTimeout(anyInt());
     verify(mockLowLevelRequest, times(2)).execute();
-    verify(mockLowLevelResponse).getStatusCode();
+    verify(mockLowLevelResponse, atLeastOnce()).getStatusCode();
     expectedLogs.verifyDebug("Request failed with IOException");
   }
 
   /** Tests that a retryable error is retried enough times. */
   @Test
   public void testRetryableErrorRetryEnoughTimes() throws IOException {
-    when(mockLowLevelRequest.execute()).thenReturn(mockLowLevelResponse);
+    List<MockLowLevelHttpResponse> responses = new ArrayList<>();
     final int retries = 10;
-    when(mockLowLevelResponse.getStatusCode())
+
+    // The underlying http library calls getStatusCode method of a response multiple times. For a
+    // response, the method should return the same value. Therefore this test cannot rely on
+    // `mockLowLevelResponse` variable that are reused across responses.
+    when(mockLowLevelRequest.execute())
         .thenAnswer(
-            new Answer<Integer>() {
+            new Answer<MockLowLevelHttpResponse>() {
               int n = 0;
 
               @Override
-              public Integer answer(InvocationOnMock invocation) {
-                return n++ < retries ? 503 : 9999;
+              public MockLowLevelHttpResponse answer(InvocationOnMock invocation) throws Throwable {
+                MockLowLevelHttpResponse response = mock(MockLowLevelHttpResponse.class);
+                responses.add(response);
+                when(response.getStatusCode()).thenReturn(n++ < retries ? 503 : 9999);
+                return response;
```

Comment:
Previously it was reusing the same mock object for different responses. Now it prepares different mock objects for different responses.